### PR TITLE
Switch to airbnb eslint

### PR DESCRIPTION
**Pull request by ayolab**

### CODE REVIEW

**Review by TheMaskedTurtle — CHANGES_REQUESTED** [review]

File: src/utils/AuthService.ts [hunk 1]
```diff
@@ -60,15 +63,204 @@ function reloadTimerOnExpiresIn(
...
 ) {
-    // TODO: Can we stop doing it in the hash for implicit flow ? To make it common for both flows
-    // Not allowed by TS because expires_in is supposed to be readonly
-    // @ts-ignore
-    user.expires_in = expiresIn;
-    userManager.storeUser(user).then(() => {
+    const updatedUser: User = {
+        ...user,
+        expires_in: expiresIn,
+        toStorageString: () => JSON.stringify(user),
```
[the remark this answers is not in the document]
Comment (ayolab) [reply src/utils/AuthService.ts:69]: The User class has `toStorageString` so I had to add it

File: src/components/inputs/react-hook-form/utils/field-label.tsx [hunk 2]
```diff
@@ -12,17 +12,22 @@ type FieldLabelProps = {
...
             <FormattedMessage id={label} values={values} />
             {optional && <FormattedMessage id="Optional" />}
         </>
     );
+}
+
+FieldLabel.defaultProps = {
+    optional: false,
+    values: undefined,
 };
```
[the remark this answers is not in the document]
Comment (ayolab) [reply src/components/inputs/react-hook-form/utils/field-label.tsx]: removed and I marked the props of the component as read-only

File: src/components/ReportViewer/report-item.tsx [hunk 3]
```diff
@@ -123,6 +124,10 @@ const ReportItem = (props: PropsWithChildren<ReportItemProps>) => {
             {...other}
         />
     );
+}
+
+ReportItem.defaultProps = {
+    className: undefined,
 };
```
Comment (TheMaskedTurtle) [inline src/components/ReportViewer/report-item.tsx:131]: Should be kept ?

File: src/components/TopBar/TopBar.tsx [hunk 4]
```diff
@@ -659,29 +643,15 @@ const TopBar = ({
...
+}
+
+TopBar.defaultProps = {
+    onParametersClick: undefined,
+    onAboutClick: undefined,
+    onThemeClick: undefined,
+    theme: undefined,
+    onEquipmentLabellingClick: undefined,
+    equipmentLabelling: undefined,
 };
```
Comment (TheMaskedTurtle) [inline src/components/TopBar/TopBar.tsx:655]: Should be kept ?

File: src/components/inputs/react-hook-form/provider/custom-form-provider.tsx [hunk 5]
```diff
@@ -39,16 +39,24 @@ const CustomFormProvider = (props: CustomFormProviderProps) => {
...
                 {children}
             </CustomFormContext.Provider>
         </FormProvider>
     );
+}
+
+CustomFormProvider.defaultProps = {
+    removeOptional: false,
+    language: undefined,
 };
```
Comment (TheMaskedTurtle) [inline src/components/inputs/react-hook-form/provider/custom-form-provider.tsx:60]: Should be kept ?

File: src/components/inputs/react-hook-form/utils/text-field-with-adornment.tsx [hunk 6]
```diff
@@ -121,18 +119,22 @@ const TextFieldWithAdornment: FunctionComponent<TextFieldWithAdornmentProps> = (
...
-            onBlur={(e) => setIsFocused(false)}
+            onFocus={() => setIsFocused(true)}
+            onBlur={() => setIsFocused(false)}
         />
     );
+}
+
+TextFieldWithAdornment.defaultProps = {
+    handleClearValue: undefined,
 };
```
Comment (TheMaskedTurtle) [inline src/components/inputs/react-hook-form/utils/text-field-with-adornment.tsx:138]: Should be kept ?

File: src/components/inputs/select-clearable.tsx [hunk 7]
```diff
@@ -53,15 +52,20 @@ const SelectClearable: FunctionComponent<SelectClearableProps> = (props) => {
...
             )}
+            options={options}
             {...otherProps}
         />
     );
+}
+
+SelectClearable.defaultProps = {
+    label: undefined,
 };
```
Comment (TheMaskedTurtle) [inline src/components/inputs/select-clearable.tsx:69]: Should be kept ?

File: src/components/ReportViewer/report-item.tsx (hunk 3 above)
Comment (Tristan-WorkGH) [reply src/components/ReportViewer/report-item.tsx:131]: To delete, it's deprecated now  
https://github.com/facebook/react/pull/16210

File: .eslintrc.json [hunk 8]
```diff
@@ -1,15 +1,27 @@
...
+  "parserOptions": {
+    "project": "./tsconfig.json"
+  },
   "ignorePatterns": [
     // node_modules is implicitly always ignored
-    "dist"
+    "dist",
+    "vite.config.mts",
+    "jest.config.ts",
+    "jest.setup.ts"
```
Comment (Tristan-WorkGH) [inline .eslintrc.json:17]: Why?

File: babel.config.json [hunk 9]
```diff
@@ -1,8 +1,12 @@
 {
     "presets": [
+        "airbnb",
         "@babel/preset-env",
         ["@babel/preset-react", { "runtime": "automatic" }],
         "@babel/preset-typescript",
         "babel-preset-vite"
```
Comment (Tristan-WorkGH) [inline babel.config.json:7]: Is the order important here?

File: .eslintrc.json [hunk 10]
```diff
@@ -1,15 +1,27 @@
...
   ],
   "rules": {
     "prettier/prettier": "warn",
-    "curly": "error"
+    "curly": "error",
+    "no-console": "off",
+    "react/jsx-props-no-spreading": "off",
+    "react/react-in-jsx-scope": "off",
```
Comment (Tristan-WorkGH) [inline .eslintrc.json:24]: I see in the doc
> If you are using the [new JSX transform from React 17](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#removing-unused-react-imports), you should disable this rule by extending [react/jsx-runtime](https://github.com/jsx-eslint/eslint-plugin-react/blob/8cf47a8ac2242ee00ea36eac4b6ae51956ba4411/index.js#L165-L179) in your eslint config (add "plugin:react/jsx-runtime" to "extends").

File: demo/data/TreeViewFinder.jsx [hunk 11]
```diff
@@ -154,20 +156,20 @@ function fetchInfinitePokemonList() {
 }
 
 function fetchInfinitePokemonTree(nodeId) {
-    IDCounter++;
+    IDCounter += 1;
```
Comment (Tristan-WorkGH) [inline demo/data/TreeViewFinder.jsx:159]: did we really want this sort of rules?

I personally prefer the `++` notation.

File: demo/src/FlatParametersTab.jsx [hunk 12]
```diff
@@ -198,7 +198,7 @@ const EXAMPLE_PARAMETERS = [
     },
 ];
 
-export const FlatParametersTab = () => {
+function FlatParametersTab() {
```
Comment (Tristan-WorkGH) [inline demo/src/FlatParametersTab.jsx:201]: function? In TS that would be
```typescript
const FlatParametersTab: FunctionComponent = () => {
```

File: demo/src/TableTab.jsx [hunk 13]
```diff
@@ -29,16 +29,17 @@ const evenThenOddOrderingKey = (n) => {
     if (n <= 0 && remainder < 1) {
         // first negative even and zero ]...-3,-2,-1]
         return n / 2 - 1;
-    } else if (n > 0 && remainder < 1) {
+    }
+    if (n > 0 && remainder < 1) {
```
Comment (Tristan-WorkGH) [inline demo/src/TableTab.jsx:33]: I don't understand this change...

File: demo/src/TreeViewFinderConfig.jsx [hunk 14]
```diff
@@ -29,7 +30,7 @@ import {
  * @param {EventListener}   onSelectionTypeChange - onChange type EventListener on the multiselect value change.
  * @param {EventListener}   onOnlyLeavesChange - onChange type EventListener on the onlyLeaves value change.
  */
-const TreeViewFinderConfig = (props) => {
+function TreeViewFinderConfig(props) {
```
Comment (Tristan-WorkGH) [inline demo/src/TreeViewFinderConfig.jsx:33]: why this change?

File: demo/src/app.jsx [hunk 15]
```diff
@@ -4,27 +4,19 @@
...
+/* eslint-disable no-nested-ternary */
+/* eslint-disable no-return-assign */
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable no-promise-executor-return */
+/* eslint-disable @typescript-eslint/no-unused-expressions */
+/* eslint-disable no-alert */
+/* eslint-disable no-undef */
+/* eslint-disable @typescript-eslint/no-shadow */
+/* eslint-disable react/jsx-no-bind */
+/* eslint-disable react/prop-types */
```
Comment (Tristan-WorkGH) [inline demo/src/app.jsx:17]: `eslint-disable` is multi-args
```suggestion
/* eslint-disable func-names no-nested-ternary no-return-assign @typescript-eslint/no-unused-vars no-promise-executor-return @typescript-eslint/no-unused-expressions no-alert no-undef @typescript-eslint/no-shadow react/jsx-no-bind react/prop-types */
```

**Review by Tristan-WorkGH — CHANGES_REQUESTED** [review]
I'll finish the review later, my navigator don't nearly freeze with 187 files to show...

File: demo/src/TreeViewFinderConfig.jsx (hunk 14 above)
Comment (ayolab) [reply demo/src/TreeViewFinderConfig.jsx:33]: https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/function-component-definition.md

File: demo/src/TableTab.jsx (hunk 13 above)
Comment (ayolab) [reply demo/src/TableTab.jsx:33]: https://eslint.org/docs/latest/rules/no-else-return

File: demo/src/FlatParametersTab.jsx (hunk 12 above)
Comment (ayolab) [reply demo/src/FlatParametersTab.jsx:201]: https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/function-component-definition.md

File: babel.config.json (hunk 9 above)
Comment (ayolab) [reply babel.config.json:7]: yes

File: .eslintrc.json (hunk 8 above)
Comment (ayolab) [reply .eslintrc.json]: There was an error during build because these files are not included in `tsconfig.json`, I created a separate tsconfig for eslint to keep the main `tsconfig.json` unchanged

File: demo/data/TreeViewFinder.jsx (hunk 11 above)
Comment (ayolab) [reply demo/data/TreeViewFinder.jsx:159]: We didn't have any discussion about it, so I kept it

File: .eslintrc.json (hunk 8 above)
Comment (ayolab) [reply .eslintrc.json:18]: Finally, I kept it as it is, so these config and setup files would be ignored

**Review by TheMaskedTurtle — APPROVED** [review]

**Review by Tristan-WorkGH — APPROVED** [review]
I'm not satisfied by some rules, but I unblock the PR.

File: src/components/Login/Login.tsx [hunk 16]
```diff
@@ -4,7 +4,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
-
+/* eslint-disable jsx-a11y/anchor-is-valid */
```
Comment (TheMaskedTurtle) [inline src/components/Login/Login.tsx:7]: Can you disable the rule just at the line level ?

File: src/components/CardErrorBoundary/card-error-boundary.tsx [hunk 17]
```diff
@@ -47,6 +47,7 @@ const ExpandMore = styled((props: ExpandMoreProps) => {
 // Extracted from https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/error_boundaries/ for types
 
 interface Props {
+    // eslint-disable-next-line react/require-default-props
     children?: ReactNode;
```
Comment (TheMaskedTurtle) [inline src/components/CardErrorBoundary/card-error-boundary.tsx:51]: Maybe this children props could be declared as required

File: src/components/filter/expert/expert-filter-utils.ts [hunk 18]
```diff
@@ -239,6 +245,7 @@ export function exportExpertRules(query: RuleGroupType): RuleGroupTypeExport {
 
         // a composite rule is a rule with dataType COMBINATOR  => build a group with child rules
         if (dataType === DataType.COMBINATOR) {
+            // eslint-disable-next-line @typescript-eslint/no-use-before-define
```
Comment (TheMaskedTurtle) [inline src/components/filter/expert/expert-filter-utils.ts:248]: This one could be fixed no ?

File: src/services/study.ts [hunk 19]
```diff
@@ -8,25 +8,18 @@
...
+        `${PREFIX_STUDY_QUERIES}/v1/studies/${studyUuid}/filters/${filterUuid}/elements`,
         {
             method: 'get',
             headers: { 'Content-Type': 'application/json' },
         },
         token
     );
 }
+
+export default exportFilter;
```
Comment (TheMaskedTurtle) [inline src/services/study.ts:25]: Given the structure, this one is a bit unexpected, maybe you can disable the rule for this one.

File: src/components/filter/expert/expert-filter-utils.ts (hunk 18 above)
Comment (ayolab) [reply src/components/filter/expert/expert-filter-utils.ts:248]: no, `transformRule` is using `transformCompositeRule` which is using `transformRule`

**Review by Tristan-WorkGH — COMMENTED** [review]
some remarks and questions

File: src/utils/AuthService.ts [hunk 20]
```diff
@@ -60,15 +63,204 @@ function reloadTimerOnExpiresIn(
     userManager: UserManager,
     expiresIn: number
 ) {
-    // TODO: Can we stop doing it in the hash for implicit flow ? To make it common for both flows
-    // Not allowed by TS because expires_in is supposed to be readonly
-    // @ts-ignore
-    user.expires_in = expiresIn;
-    userManager.storeUser(user).then(() => {
+    const updatedUser: User = {
```
Comment (TheMaskedTurtle) [inline src/utils/AuthService.ts:66]: I tested a bit it feels like it breaks the automatic token renewal with custom values... This part of the code is quite touchy it would be nicer it you don't change code and just disable eslint rules here

**Review by TheMaskedTurtle — CHANGES_REQUESTED** [review]

File: src/components/MuiVirtualizedTable/KeyedColumnsRowIndexer.tsx [hunk 21]
```diff
@@ -128,35 +146,252 @@ export interface FilteredRows {
...
+        return 1;
+    }
+    if (fuzzySign === 'asc') {
+        return 1;
+    }
+    if (fuzzySign === 'desc') {
+        return -1;
+    }
+    return 0;
+};
```
[the remark this answers is not in the document]
Comment (TheMaskedTurtle) [reply src/components/MuiVirtualizedTable/KeyedColumnsRowIndexer.tsx:165]: There are already a lot of changes we should not modify the code too much in this PR.

[the remark this answers is not in the document]
Comment (ayolab) [reply src/components/MuiVirtualizedTable/KeyedColumnsRowIndexer.tsx:164]: anyway, it will soon be removed

File: src/components/TopBar/TopBar.tsx [hunk 22]
```diff
@@ -362,15 +360,15 @@ const TopBar = ({
                             onClick={handleToggleSettingsMenu}
                             color="inherit"
                             style={
-                                Boolean(anchorElSettingsMenu)
+                                anchorElSettingsMenu
```
[the remark this answers is not in the document]
Comment (ayolab) [reply src/components/TopBar/TopBar.tsx:363]: yes

**Review by TheMaskedTurtle — APPROVED** [review]

File: src/components/DirectoryItemSelector/directory-item-selector.tsx [hunk 23]
```diff
@@ -223,7 +349,7 @@ const DirectoryItemSelector: FunctionComponent<DirectoryItemSelectorProps> = ({
     return (
         <TreeViewFinder
             onTreeBrowse={fetchDirectory as (NodeId: string) => void}
-            sortMethod={sortHandlingDirectories}
+            sortMethod={(a, b) => sortHandlingDirectories(a, b)}
```
[the remark this answers is not in the document]
Comment (Tristan-WorkGH) [reply src/components/DirectoryItemSelector/directory-item-selector.tsx:352]: and it's exactly what the rule don't want (an arrow function as jsx props): https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md

[the remark this answers is not in the document]
Comment (TheMaskedTurtle) [reply src/components/DirectoryItemSelector/directory-item-selector.tsx:352]: @ayolab I think you can just move this method out of the component, so we fix this problem

File: src/components/TopBar/AboutDialog.tsx [hunk 24]
```diff
@@ -388,21 +537,20 @@ const AboutDialog = ({
                                         <>
                                             {[...modules]
                                                 .sort(compareModules)
-                                                .map((module, idx) => (
+                                                .map((module) => (
                                                     <Module
-                                                        key={`module-${idx}`}
+                                                        key={`module-${module.name}`}
```
[the remark this answers is not in the document]
Comment (ayolab) [reply src/components/TopBar/AboutDialog.tsx]: Done

File: src/components/DirectoryItemSelector/directory-item-selector.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (ayolab) [reply src/components/DirectoryItemSelector/directory-item-selector.tsx]: Done

File: src/components/ElementSearchDialog/tag-renderer.tsx [hunk 25]
```diff
@@ -41,7 +41,7 @@ function TagRenderer({ element, ...props }: TagRendererProps) {
             />
         );
     }
-    return null;
+    return undefined;
```
Comment (Tristan-WorkGH) [inline src/components/ElementSearchDialog/tag-renderer.tsx:44]: just don't return then
```suggestion
```

File: src/components/TopBar/AboutDialog.tsx [hunk 26]
```diff
@@ -539,9 +539,10 @@ function AboutDialog({
                                         <>
                                             {[...modules]
                                                 .sort(compareModules)
-                                                .map((module) => (
+                                                .map((module, idx) => (
                                                     <Module
-                                                        key={`module-${module.name}`}
+                                                        // eslint-disable-next-line react/no-array-index-key
```
Comment (Tristan-WorkGH) [inline src/components/TopBar/AboutDialog.tsx:544]:
```suggestion
                                                   // eslint-disable-next-line react/no-array-index-key -- modules aren't necessary uniques in the API, so we need something else for the key that must be unique
```

**Review by Tristan-WorkGH — APPROVED** [review]
